core/src/main/java/io/grpc/internal/TransportSet2.java

```diff
@@ -0,0 +1,501 @@
+      this.address = address;
+    }
+
+    @Override
+    public void transportReady() {
+      if (log.isLoggable(Level.FINE)) {
+        log.log(Level.FINE, "[{0}] {1} for {2} is ready",
+            new Object[] {getLogId(), transport.getLogId(), address});
+      }
+      ConnectivityState savedState;
+      synchronized (lock) {
+        savedState = state.getState();
+        reconnectPolicy = null;
+        nextAddressIndex = 0;
+        if (savedState == SHUTDOWN) {
+          // activeTransport should have already been set to null by shutdown(). We keep it null.
+          Preconditions.checkState(activeTransport == null,
+              "Unexpected non-null activeTransport");
+        } else if (pendingTransport == transport) {
+          gotoNonErrorState(READY);
+          activeTransport = transport;
+          pendingTransport = null;
+        }
+      }
+      if (savedState == SHUTDOWN) {
+        transport.shutdown();
+      }
+    }
+
+    @Override
+    public void transportInUse(boolean inUse) {
+      handleTransportInUseState(transport, inUse);
+    }
+
+    @Override
+    public void transportShutdown(Status s) {
+      boolean allAddressesFailed = false;
+      if (log.isLoggable(Level.FINE)) {
+        log.log(Level.FINE, "[{0}] {1} for {2} is being shutdown with status {3}",
+            new Object[] {getLogId(), transport.getLogId(), address, s});
+      }
+      Runnable runnable = null;
+      synchronized (lock) {
+        if (state.getState() == SHUTDOWN) {
+          return;
+        }
+        if (activeTransport == transport) {
+          gotoNonErrorState(IDLE);
+          activeTransport = null;
+        } else if (pendingTransport == transport) {
+          // Continue reconnect if there are still addresses to try.
+          if (nextAddressIndex == 0) {
+            allAddressesFailed = true;
+          } else {
+            Preconditions.checkState(state.getState() == CONNECTING,
+                "Expected state is CONNECTING, actual state is %s", state.getState());
+            runnable = startNewTransport();
+          }
+        }
+      }
+      if (allAddressesFailed) {
+        // Initiate backoff
+        // Transition to TRANSIENT_FAILURE
+        scheduleBackoff(s);
+      }
+      if (runnable != null) {
+        runnable.run();
+      }
+    }
+
+    @Override
+    public void transportTerminated() {
+      if (log.isLoggable(Level.FINE)) {
+        log.log(Level.FINE, "[{0}] {1} for {2} is terminated",
+            new Object[] {getLogId(), transport.getLogId(), address});
+      }
+      boolean terminated = false;
+      handleTransportInUseState(transport, false);
+      synchronized (lock) {
+        transports.remove(transport);
+        if (state.getState() == SHUTDOWN && transports.isEmpty()) {
+          if (log.isLoggable(Level.FINE)) {
+            log.log(Level.FINE, "[{0}] Terminated in transportTerminated()", getLogId());
+          }
+          handleTermination();
```

Comment:
TransportSet is terminated when both of the two conditions have been met: 1) `shutdown()` has been called; 2) all transports have been terminated. When a transport is shutdown, it just stops accepting new streams, but doesn't kill existing streams. It waits until all existing streams to complete to become terminated.

If this line is reached, it means when shutdown() was called, there were still not-yet terminated transports, thus shutdown() wouldn't call handleTermination(), so we need to call handleTermination().